### ISSUE TITLE
Protected version duk_json_decode()

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#386]
There should be a protected version of duk_json_decode() which doesn't throw an error on invalid input but instead leaves the error value on top of the stack (and perhaps change it to return a duk_bool_t indicating success or failure).  Currently this requires either wrapping the JSON decode in a `duk_safe_call()` or `pcall`ing `JSON.decode()` directly from C, both of which are inconvenient when all you need is a quick JSON decode unrelated to normal script execution.  The latter may even break if an evil script replaces the `JSON` object!

**Comment by svaarala** [comment]
I understand what you mean - but should every unsafe call in the API have a protected version for consistency? If not, why is `duk_json_decode()` special?

**Comment by fatcerberus** [comment]
For the same reason `duk_call` and `duk_eval` have protected variants, usually you're calling into Duktape with user input (in call and eval's case, code, in JSON's case, a JSON literal) and so errors are expected--having it throw is inconvenient.

My particular case is that I'm using Duktape to parse source maps (since I'm already dependent on Duktape, I saw no reason not to leverage it as a JSON decoder :) ) something like this:

```
const char*
sfs_sourcemap_path(sandbox_t* fs, const char* path)
{
	static char retval[SPHERE_PATH_MAX];

	if (fs->sourcemap == NULL)
		retval[0] = '\0';
	else {
		duk_push_lstring_t(g_duk, fs->sourcemap);
		duk_json_decode(g_duk, -1);  // if this throws, I'm dead
		duk_pop(g_duk);
		// work with the object here
	}
	return retval;
}
```

I could wrap the actual `duk_json_decode()` in a safecall, but that feels dirty to me, it would be much cleaner if I could handle the failure inline.

For other APIs that can throw, you're usually not dealing directly with user input (or it's already been sanitized), so a thrown error is exceptional and the lack of a protected variant is fine.

**Comment by svaarala** [comment]
How is this different from `duk_get_prop()` from a random place? It's calling into Duktape with potential user input (say a configuration key to look up)?

**Comment by fatcerberus** [comment]
`duk_get_prop()` isn't a perfect analogy though--it won't itself throw (unless you give it a bad index or something that's not object coercible, both of which would be bugs), there are functions to check whether the value retrieved is of the proper type (duk_is_string() et al.)--and get_prop() itself returns a bool to indicate whether the lookup succeeded.  With JSON, there's no way to validate the input at all.

**Comment by svaarala** [comment]
`duk_get_prop()` can quite easily throw on getters for example.

**Comment by svaarala** [comment]
Please note that I *do* understand it's inconvenient to protected call wrap the decode in your case. I'm just trying to figure out what a consistent policy would be so that I won't end up adding hundreds of protected calls to the API; someone always marginally benefits from a new protected call, but adding them all means a measurable impact on footprint as safe calls can't be implemented as pure macros.

That's why I'm trying to have a consistent line in what protected variants are protected. So far: compilation/eval, Duktape function calls, and native function calls. The only real deviation this so far is `duk_safe_to_string()` which was so universally needed for dumping results that I caved in... :)

**Comment by fatcerberus** [comment]
I guess my biggest issue is that, with the current API, there's no way to trial-parse JSON easily, you always need to set up a safe call harness if you don't want your C program to blow up due to invalid JSON.  Even if the decode itself wasn't protected, a `duk_is_valid_json()` function would suffice--but though would probably be even more bloat.

An alternative to adding a bunch of protected functions would be to update the safe call method so that, instead of the way it works now (which I always found to be a bit awkward), there could be a function you would call to set up a catcher (essentially, a `setjmp` analogue), and if any subsequent call throws, control returns to the catcher.  The downside would be that the native code would have to remember to tear down that catcher when it was no longer needed (longjmp'ing into a function which already returned = bad).

**Comment by svaarala** [comment]
I know, but note that a failing getter would equally fail on `duk_get_prop()` and there's no way to validate against that. It might happen less often but it's still quite usual when getters encounter unexpected inputs.

Similarly you can encounter an out-of-memory or an internal error in any location - and unless you're running in a protected wrapper call the error will be fatal. If you really want you program to be reliable, you'll need to wrap everything except protected calls for this reason.

**Comment by svaarala** [comment]
I whole-heartedly agree that a *convenient* way of making arbitrary calls or statement blocks proteced in C code would be very welcome. So far there haven't been too many good ideas for that :(

Technically one could do something like:

```
DUK_SAFE({
    /* block of user code */
}, {
    /* success handling */
}, {
    /* error handling */
});
```

The macro would turn this into a SETJMP structure like:

```
if (setjmp(...) != 0) {
    duk__handle_longjmp(...); /* unwind etc */

    /* error handling */

    duk_pop(ctx);  /* pop error */
} else {
    /* block of user code */

    /* success handling code */

    duk_pop(ctx);  /* pop error */
}
```

The particular setjmp function/macro etc would of course be hidden from user code; it'd just rely on that macro.

This is not particularly nice either, but at least the statement blocks would see the external scope unlike with `duk_safe_call()`.

**Comment by svaarala** [comment]
Hmm that `/* success handling */` part in the macro example would be unnecessary so it'd look something like:

```
DUK_SAFE({
    /* block of user code */
}, {
    /* error handling */
});
```

Value stack policy would be interesting: if it were consistent and value stack was restored after the safe blocks, you'd have the problem that anything pushed on the stack in the "block of user code" was lost. In particular, if you did `duk_push_something(); duk_json_encode(ctx, -1); my_str = duk_require_string(ctx, -1)`, that `my_str` would become dangling when the block ended.

**Comment by fatcerberus** [comment]
I think a good metric for determining when a protected call is needed would be to ask the question, "When this API throws an error, what is the likelihood that this is exceptional?"  If failure is not exceptional (i.e. is expected during normal operation), a protected call would probably be welcome.

I fully understand the hesitance though, bloat is bad. :)

**Comment by svaarala** [comment]
That's a pretty good basic rule I think. But people have a widely different concept of what's exceptional :)

For example, I'd find it odd if one could cause a program to encounter a fatal error from a getter call (or any "normal" script code for that matter). That's especially the case if you use Duktape to run user plugins whose code quality may vary a great deal; in that situation I'd consider even forever loops (script timeouts) "normal" and to be expected.

Someone else might think you just fix the getter and a fatal error is just fine. In fact there are legitimate cases where protected call wrapping is never needed: you just crash and restart and it's sometimes fine.

My view is probably motivated by working with embedded systems and having the need to diagnose errors from the field (even if, or perhaps especially, if the errors are caused by script errors). In that situation you want to catch errors, write out diagnostics, and deliver that diagnostics for analysis e.g. on the next device reboot.

**Comment by fatcerberus** [comment]
Yeah, reworking the safe call API would be a better solution in the long run than getting stuck with a bunch of `duk_p*` variants that you can't remove without breaking SemVer guarantees.  The biggest issue with the safe call API right now, as you pointed out above, is that unlike try-catch from script, the local scope is inaccessible during the protected section.  If that weren't the case right now, I'd have no issue with `duk_json_decode()` throwing. :)

**Comment by fatcerberus** [comment]
A downside to the try-catch style macro idea above, is that since it is ultimately a setjmp() under the hood, all the usual caveats apply re: non-volatile variables, signal mask, etc. which might not be intuitive.

So yeah, not sure what the right answer is here.

**Comment by fatcerberus** [comment]
Anyway, for JSON decoding in particular (the original topic of this issue), I can see a scenario where an existing Duktape dependency is leveraged in lieu of a more lightweight JSON decoder because, well, it's already there so may as well use it.  But with the current setup, it's not optimal as a replacement for a lightweight decoder because it can throw.  So I think that makes it a unique case compared to all the other throwing APIs.

**Comment by fatcerberus** [comment]
I guess my argument here is that `duk_json_decode()` is more likely to be called "outside" of normal script processing, in which case if the error escapes into script there will be no context to deal with it.  As in my case, the JSON input may not even have originated from JS code.

**Comment by fatcerberus** [comment]
If that was still confusing, read "likely to be called outside normal script execution" like "called as a normal C function". :)

**Comment by mpietras** [comment]
For the record I ran into this as well... albeit 3 years later.  I have JSON results from a server I want to parse in c++ for simply leaving on the duk stack as a return object from a native-implemented method but I can't guarantee the JSON is valid before parsing...

**Comment by svaarala** [comment]
The easiest way to deal with that right now is something like:
```c
static duk_ret_t try_parse(duk_context *ctx) {
    duk_json_decode(ctx, -1);
    return 1;
}

/* ... actual call site ... */

duk_push_string(ctx, json_source);
if (duk_safe_call(ctx, try_parse, NULL, 1 /*nargs*/, 1 /*nrets*/) == DUK_EXEC_SUCCESS) {
    /* Parse successful, object or other decoded value at stack top. */
} else {
    /* Error, error object at stack top. */
}
duk_pop(ctx);
```

So, it's not ideal, but it's still quite understandable. Also the pattern holds over any other unprotected API calls.

**Comment by mpietras** [comment]
I was looking into that but was unclear on the exact usage - thank you for the example.  Not pretty as you said, but works fine.
